File: binary_tree_level_order_traversal.py

class Node(object):
    def __init__(self,val):
        self.val = val
        self.left = None
        self.right = None


class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def levelOrder(self, root):


if __name__ == "__main__":
    n3 = Node(3)
    n9 = Node(9)
    n20 = Node(20)
    n15 = Node(15)
    n7 = Node(7)
    n3.left = n9
    n2.right = n20
    n20.left = n15
    n20.right = n7
    print Solution().levelOrder(n3)

File: binary_tree_level_order_traversal.cpp
/*
Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

For example:
Given binary tree {3,9,20,#,#,15,7},

    3
   / \
  9  20
    /  \
   15   7


return its level order traversal as:

[
  [3],
  [9,20],
  [15,7]
]
*/

#include "header.h"

vector<vector<int> > levelOrder(TreeNode* root) {
    vector<vector<int> > res;
    if(!root) return res;
    queue<TreeNode*> qs[2];
    qs[0].push(root);
    int curr = 0;
    vector<int> temp;
    while(!qs[curr%2].empty()){
        TreeNode* t = qs[curr%2].front();
        qs[curr%2].pop();
        temp.push_back(t->val);
        if(t->left) qs[(curr+1)%2].push(t->left);
        if(t->right) qs[(curr+1)%2].push(t->right);
        if(qs[curr%2].empty()){
            curr++;
            res.push_back(temp);
            temp.clear();
        }
    }
    return res;
}

int main(){
    vector<string> arr = {"3","9","20","#","#","15","7"};
    TreeNode* root = create_tree(arr);
    vector<vector<int> > res = levelOrder(root);
    for(auto i:res) print_vector<int>(i);
    return 0;
}